### PR TITLE
Revised update process

### PR DESCRIPTION
**Pull request by mokaddem**
## Fix
- Previously, ``database_number`` where incremented if the update process was locked
- Do not increment ``database_number`` if ``flag_stop`` is set and errors occurs
- Added sidebar
- Fixed FA classes

## New
- ``SQLSTATE[42S21]: Column already exists: 1060 Duplicate column name`` and ``SQLSTATE[42000]: Syntax error or access violation: 1061 Duplicate key name`` are not considered as an error anymore
- Changed worker place in the update process
- When in cake console, perform the update inline (instead of creating a new worker)
- Deleted 1 useless function
- Created a new worker named `update` which take care of all update processes
- Added database schema diagnostic in the `diagnostic` page 
- Added *TaskScheduler* library
- Improved ``updateProgress`` page
- Added another lock system that triggers if an update fails multiple time in a row. Therefore, preventing infinite update loop

## Note
Regarding question https://github.com/MISP/MISP/pull/4605#issuecomment-490865669
It should not be an issue as updates are executed only if the lock is not locked.

In the scenario from the question:
- You trigger the update
- The update is delegated to a worker
- You restart/kill the apache-server/worker
    - The ``database_number`` is not incremented
    - The lock is still locked
- No new updates can take place until the lock unlocked after a timer (default timer = 2h; can be changed)
- Eventually, updates are executed again; complete; ``database_number`` gets incremented

### CODE REVIEW

File: app/Model/AppModel.php [hunk 1]
```diff
@@ -106,14 +106,15 @@ public function afterSave($created, $options = array())
     // this could become useful in the future
     public function updateMISP($command)
     {
+        $db_update_success = false;
         switch ($command) {
             case '2.4.20':
-                $this->updateDatabase($command);
+                $db_update_success = $this->updateDatabase($command);
```
Comment (JakubOnderka) [inline app/Model/AppModel.php]: I think variable names should be `camelCase` in MISP.

Comment (mokaddem) [reply app/Model/AppModel.php]: Hey @JakubOnderka thanks for the feedback (I just saw it..)!
Indeed, ``camelCase`` is widely used in the MISP codebase.
I must admit that I used the ``under_score`` notation because of personal taste but also because the rule is not applied everywhere (there are mixes between ``cameCase`` and ``under_score`` in a lot of places).

Even though (AFAIR) PSR-x does not explicitly specify the notation to be used on variable, I agree that having an homogeneous  code base would be better.

What do you think?

Comment (iglocska) [reply app/Model/AppModel.php]: Indeed, there's no recommendation besides being consistent. I am not too bothered with it either way, though it would look nicer if we were more consistent on it.

File: app/Console/Command/AdminShell.php [hunk 2]
```diff
@@ -550,4 +525,25 @@ public function purgeFeedEvents()
             }
         }
     }
+
+    public function dumpCurrentDatabaseSchema()
```
Comment (iglocska) [inline app/Console/Command/AdminShell.php:523]: Nice idea

File: app/Model/AppModel.php [hunk 3]
```diff
@@ -236,51 +238,41 @@ private function __addNewFeeds($feeds)
...
         if ($this->isUpdateLocked()) {
+            $this->Log->create();
+            $this->Log->save(array(
+                    'org' => 'SYSTEM',
+                    'model' => 'Server',
+                    'model_id' => 0,
+                    'email' => 'SYSTEM',
+                    'action' => 'update_database',
+                    'user_id' => 0,
+                    'title' => __('Issues executing the SQL query for ') . $command,
```
Comment (iglocska) [inline app/Model/AppModel.php]: really tiny improvement - change this to:
```
 'title' => __('Issues executing the SQL query for %s', $command),
```

That will allow translators to deal with languages with different grammar structures where you would have a different sentence order.

I realise that this has been there from before, but perhaps a good moment to fix it ;)

File: app/Model/AppModel.php [hunk 4]
```diff
@@ -1340,43 +1332,54 @@ public function updateDatabase($command, $useWorker=false)
...
-                    $this->__setUpdateError($i);
-                    $error_count++;
-                    if ($exitOnError) {
-                        $flag_stop = true;
-                        break;
+                        'change' => __('The executed SQL query was: ') . $sql . PHP_EOL . __(' The returned error is: ') . $error_message
+                    );
+                    $this->__setUpdateResMessages($i, __('Issues executing the SQL query for ') . $command . __('. The returned error is: ') . PHP_EOL . $error_message);
+                    $error_duplicate_column = 'SQLSTATE[42S21]: Column already exists: 1060 Duplicate column name';
+                    $error_duplicate_index = 'SQLSTATE[42000]: Syntax error or access violation: 1061 Duplicate key name';
```
Comment (iglocska) [inline app/Model/AppModel.php]: The only issue I can see with this sort of error handling is that we're tying ourselves even closer to mysql. Perhaps something more generic would be handy? Simple substring search for duplicate key name / duplicate column name?

File: app/Model/Server.php [hunk 5]
```diff
@@ -4255,6 +4255,140 @@ public function redisInfo()
         return $output;
     }
 
+    public function dbSchemaDiagnostic()
```
Comment (iglocska) [inline app/Model/Server.php:4266]: Cool stuff.

File: app/View/Elements/genericElements/IndexTable/headers.ctp [hunk 6]
```diff
@@ -31,5 +31,8 @@
             __('Actions')
         );
     }
-    echo $headersHtml;
+    $thead = '<thead>';
```
Comment (iglocska) [inline app/View/Elements/genericElements/IndexTable/headers.ctp:34]: For next time if possible: Have generic changes like these in a separate commit, makes reviewing easier (don't change it for now though!)

File: app/View/Elements/genericElements/IndexTable/index_table.ctp [hunk 7]
```diff
@@ -24,8 +24,8 @@
     if (!empty($data['html'])) {
         echo sprintf('<p>%s</p>', $data['html']);
     }
-    $skipPagination = isset($data['skip_pagination']) && $data['skip_pagination'];
-    if (!$skipPagination) {
+    $skip_pagination = isset($data['skip_pagination']) ? $data['skip_pagination'] : 0;
```
Comment (iglocska) [inline app/View/Elements/genericElements/IndexTable/index_table.ctp]: this was fixed in a different PR since then, should also not really be part of this PR.

**Review by iglocska — COMMENTED** [review]
Looks great, go for it!

File: app/Model/AppModel.php (hunk 4 above)
Comment (mokaddem) [reply app/Model/AppModel.php]: Just had a look at [mysql-error-reference](https://dev.mysql.com/doc/refman/8.0/en/server-error-reference.html). Doing a substring search for ``Duplicate column name`` works but ``Duplicate key`` is also present in other *mysql* error messages.
Moreover for *posgresql*, the error message for duplicate column is ``ERROR: column "x" specified more than once``, so a substring might not work.
What I can do is also collect and whitelist errors from *posgresql*.
What do you think?

File: app/View/Elements/genericElements/IndexTable/headers.ctp (hunk 6 above)
Comment (mokaddem) [reply app/View/Elements/genericElements/IndexTable/headers.ctp:34]: Sure! The problem was that I fixed the indexTable in this branch but in the meanwhile, it was also fixed in 2.4..
Sorry for that ;)

**Comment by mokaddem** [comment]
Switched to camelCase as requested by @JakubOnderka ;)
